### DISCUSSION TITLE
What about concurrency and parallelism?

### DISCUSSION THREAD

**Discussion by amano-kenji** [mattwparas/steel#147] [Q&A] [answered]
How will steel scheme support them?

Janet has fibers for concurrency and threads for parallelism. It passes queued messages between fibers and threads through channels.

I personally found janet's constructs for concurrency and parallelism easier and more performant and more granular than raku's.

**Comment by amano-kenji** [comment]
I think janet event loop is a very good way of achieving concurrent IO operations.

For CPU intensive tasks, there is no alternative to threads. Threads can be good if they don't cause problems. Things like software transactional memory and persistent data structure can help to write thread-safe programs.

**Reply by mattwparas** [reply]
With the current implementation of steel, most values are not thread safe. When constructing native threads, the environment is deep cloned and replicated in the new thread. This would mean you don't want to constantly spawn new threads, but rather maybe have some long lived threads available to reuse.

The VM does not have an event loop - it is inherently blocking. However, I'm not ruling out including an event loop of some kind in the future. The VM also _does_ support native async functions, and an event loop can be implemented in user space to handle these, see this for an example: https://github.com/mattwparas/steel/blob/master/examples/scripts/async-threads.scm.

**Reply by amano-kenji** [reply]
So, there is no way to run concurrent IO operations in the background, yet?

I'm not familiar with scheme. I have never used scheme before.

**Reply by mattwparas** [reply, marked as answer]
The linked file is an example of running concurrent IO operations. However, there is not a built in event loop, so the async operations are handled via an async/await style, where you yield on blocking operations to the executor which is implemented in user space.

I have not made it a priority to implement non blocking (async) operations, however I will most likely include async file system operations. You can create async functions in dylibs as well, so for example you could have a non blocking web requests library.
